### PR TITLE
Switched to AsyncKeyedLock

### PR DESCRIPTION
**Pull request by MarkCiliaVincenti**
This PR will reduce memory allocations and memory usage.

### CODE REVIEW

**Comment by MarkCiliaVincenti** [comment]
@openbullet :)

**Comment by openbullet** [comment]
Hey sorry I didn't have time to look into this yet, will do soon ^^'

**Comment by MarkCiliaVincenti** [comment]
> Hey sorry I didn't have time to look into this yet, will do soon ^^'

gentle reminder :)

**Comment by meinname** [comment]
@openbullet asked me to try it.
And i noticed that jobs don't start as they don't load proxies. - without proxies it seems to be ok. (tho i did no extensive test without being able to use proxies)

**Comment by MarkCiliaVincenti** [comment]
> @openbullet asked me to try it. And i noticed that jobs don't start as they don't load proxies. - without proxies it seems to be ok. (tho i did no extensive test without being able to use proxies) (Tested on native)

Hi,

What would be the easiest way to reproduce this issue please so that I could maybe debug? Or would you kindly be able to debug and tell me where the issue lies please?

**Comment by MarkCiliaVincenti** [comment]
@meinname

**Comment by meinname** [comment]
Simply try to start a job with proxies.
It wont start

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by MarkCiliaVincenti** [comment]
@meinname I'm going to need more info to replicate this issue please. Also can you try again please?

**Comment by MarkCiliaVincenti** [comment]
@meinname please let me know. I've also updated to the latest version of the library with performance improvements.

**Comment by meinname** [comment]
Still. When starting a Job with proxies off it seems to work.
But as soon as i use proxies (tested from OB2 Proxy Manager and from File) a Job with a simple config only remains at Starting...
(only tested this time on native Client on Windows)

**Comment by MarkCiliaVincenti** [comment]
Would you be so kind as to dumb it down into steps with screenshots to reproduce? And can you confirm that without the PR it does not give an issue?

**Comment by meinname** [comment]
I don't know hot to dumb it down more - but i try

First i check out ob2 master to a new directory and publish it to a new directory.
Then i start the native client and create a new config - basically anything you want. - i just used one as simple as that one: 

```
BLOCK:RandomInteger
  => VAR @randomIntegerOutput
ENDBLOCK

BLOCK:Keycheck
  KEYCHAIN SUCCESS OR
    INTKEY @randomIntegerOutput GreaterThanOrEqualTo 9
  KEYCHAIN FAIL OR
    INTKEY @randomIntegerOutput LessThan 9
ENDBLOCK

```

and add a few simple proxies on ob2 proxy tab to a new group.
As they aren't used, i just add some like 127.0.0.1:9000  127.0.0.1:9001 .... or real ones or whatever.

I create a new multirun job and 
select the config. - set bots to 10 - proxy mode to on - wordlist either use one via file or just use infinite
And then go to the job and start it.
It works. - job runs and you get successes and fails 

then do the same but with the PR merged
And then you can see that after pressing start ob2 does nothing. Beside job status is and stays on "Starting..."

**Comment by Gauvino** [comment]
@MarkCiliaVincenti Hello, first thank for your contribution, could please resolve the merge conflict to get this reviewed ?

**Comment by MarkCiliaVincenti** [comment]
I took a look at the updated code and I tried my best to see how I could get the AsyncKeyedLock library to work with OpenBullet2 but I don't understand what that LoliCodeBlockInstance is.

More importantly for you guys is that the current AsyncLocker class is not just very inefficient but worse than that has obvious race conditions.

You will need to refactor your code.

**Comment by openbullet** [comment]
Ok so, after like 2 years, I had some time to look into this.

Our `AsyncLocker` imple now uses `ConcurrentDictionary.GetOrAdd` (on the `staging` branch at least) so there should be no more races. I don't see a reason to justify the lib addition other than syntax sugar tbh.

The problem that your PR caused to proxies is that it wraps calls to `ProxyPool.ReloadAllAsync` in `AsyncLocker.LockAsync(typeof(ProxyPool), nameof(ReloadAllAsync))`, but `ProxyPool.ReloadAllAsync` also locks the same key. Since the lock is not reentrant, jobs with proxies can deadlock/stay at `Starting`.

Thank you anyways @MarkCiliaVincenti and sorry for the wait but I had kinda stopped working on this project 😅 wish you all the best.
